main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1264,14 +1264,24 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
           */
         def inferSelectRule(sr0: ResolvedAst.SelectChannelRule): InferMonad[(List[Ast.TypeConstraint], Type, Type)] =
           sr0 match {
-            case ResolvedAst.SelectChannelRule(sym, chan, body) => for {
+            case ResolvedAst.SelectChannelRule.SelectGet(sym, chan, body) => for {
               (chanConstrs, chanType, _) <- visitExp(chan)
               (bodyConstrs, bodyType, _) <- visitExp(body)
               _ <- unifyTypeM(chanType, Type.mkChannel(sym.tvar, sym.loc), sym.loc)
               resultCon = chanConstrs ++ bodyConstrs
               resultTyp = bodyType
               resultEff = Type.Impure
             } yield (resultCon, resultTyp, resultEff)
+
+            case ResolvedAst.SelectChannelRule.SelectPut(chan, value, body) => for {
+              (chanConstrs, chanType, _) <- visitExp(chan)
+              (valueConstrs, valueType, _) <- visitExp(value)
+              (bodyConstrs, bodyType, _) <- visitExp(body)
+              // TODO: Should I unify something here?
+              resultCon = chanConstrs ++ valueConstrs ++ bodyConstrs
+              resultTyp = bodyType
+              resultEff = Type.Impure
```

Comment:
You must also add a unification that checks that *the type of the value matches the type of the channel*.

E.g. we dont want someone to send Int32 on a String channel.

You could say `mkChannel(valueType)` and unify that with `chanType`.

or something like that. Try it out with some tests!